Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,504 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2018-2019 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import NIO
+import NIOConcurrencyHelpers
+import NIOHTTP1
+import NIOTLS
+
+/// A connection pool that manages and creates new connections to hosts respecting the specified preferences
+class ConnectionPool {
+    /// The default look group used to schedule async work
+    let loopGroup: EventLoopGroup
+
+    /// The configuration used to bootstrap new HTTP connections
+    let configuration: HTTPClient.Configuration
+
+    /// The main data structure used by the `ConnectionPool` to retreive and create connections associated
+    /// to a a given `Key` .
+    /// - Warning: This property shouldn't be directly accessed, use the `ConnectionPool` subscript instead
+    var _connectionProviders: [Key: ConnectionProvider] = [:]
+
+    private let lock = Lock()
+
+    init(group: EventLoopGroup, configuration: HTTPClient.Configuration) {
+        self.loopGroup = group
+        self.configuration = configuration
+    }
+
+    /// This enables the connection pool to store and retreive `ConnectionProvider`s
+    /// by ensuring thread safety using the `lock`.
+    private subscript(key: Key) -> ConnectionProvider? {
+        get {
+            return self.lock.withLock {
+                _connectionProviders[key]
+            }
+        }
+        set {
+            self.lock.withLock {
+                _connectionProviders[key] = newValue
+            }
+        }
+    }
+
+    /// Determines what action should be taken regarding the the `ConnectionProvider`
+    ///
+    /// This method ensures there is no race condition if called concurently from multiple threads
+    private func getAction(for key: Key, eventLoop: EventLoop) -> ProviderGetAction {
```

Comment:
This function has only one caller, which calls it with `self.loopGroup.next()`. That suggests we don't need this `eventLoop` parameter here at all, and this function can just subsume the event loop choice.

Reply:
Actually, I think this doesn't need an event loop at all. The event loop is used only to construct a promise, but that promise exists only to have a single callback attached to it. I don't think we need the flexibility here, so `.future` could instead just provide a closure that is invoked, and avoid some unnecessary bookkeeping and allocations.

Reply:
> Actually, I think this doesn't need an event loop at all. The event loop is used only to construct a promise, but that promise exists only to have a single callback attached to it.

Actually, the `Provider` type adds multiple callbacks, for instance [here](https://github.com/swift-server/async-http-client/pull/105/files#diff-d4fe528099e126b77e626c998dfbf926R199).

Reply:
An issue I encountered is that `Task`, which is returned synchronously, has an `eventLoop` property, this is the `eventLoop` the delegate is called on, but sometimes it is not possible to determine synchronously what would be the best value. The reason why is that when we have reached the maximum number of concurrent connections and a request has to wait for a connection to be released, we don't know upfront what the `EventLoop` of the next released connection will be, if we knew it or if we avoided altogether to synchronously set the `EventLoop` property of `Task` it could avoid some hoping overhead.

The https://github.com/swift-server/async-http-client/pull/105/commits/d00e3045eade2650f005bebb4da5011a1eb6be70 attempts to reduce this issue (when the `EventLoopPreference` is `.indifferent`) by using the same `EventLoop` by default for delegate/channel of requests to a certain (host, scheme, port) when the user didn't specify anything, this means when the user uses `.indifferent` only they'll get the minimum overhead.